### DISCUSSION TITLE
Enhance logging in fybrik

### DISCUSSION THREAD

**Discussion by lappanna** [fybrik/fybrik#851] [Developers]
we create fybrik application with a unique name like `fybrik-<uuid>` as shown below: 

And we do set the uuid in label `app.fybrik.io/dataAccessModuleNamespace`, but this might change in future.  

please do let us know if you want another label to be added in fybrikapplication with this uuid, so that all the  underlying fybrik layers can have this uuid in logs related to this fybrikapplication.

```
oc get fybrikapplication -n fybrik-ashwini-app fybrik-90c91fca-bda4-4c05-86c3-92c90418cb87 -o yaml
apiVersion: app.fybrik.io/v1alpha1
kind: FybrikApplication
metadata:
  creationTimestamp: "2021-10-13T05:08:27Z"
  finalizers:
  - FybrikApplication.finalizer
  generation: 3
  labels:
    app.fybrik.io/dataAccessModuleNamespace: datafabric-90c91fca-bda4-4c05-86c3-92c90418cb87
  managedFields:
  ```

**Comment by roee88** [comment]
Can't the namespace and name be the UUID?

**Reply by lappanna** [reply]
we have added another label with name just appId, it will have just UUID
```
Lathas-MacBook-Pro:fybrik lathaappanna$ oc describe fybrikapplication  fybrik-0645563a-34c7-4f65-a095-8b702ba1b26e -n fybrik-system-app 
Name:         fybrik-0645563a-34c7-4f65-a095-8b702ba1b26e
Namespace:    fybrik-system-app
Labels:       app.fybrik.io/dataAccessModuleNamespace=datafabric-0645563a-34c7-4f65-a095-8b702ba1b26e
              appId=0645563a-34c7-4f65-a095-8b702ba1b26e
```

**Reply by roee88** [reply]
OK I guess we can start by defining and documenting a special label that if present is used in logging (otherwise it will say n/a). If we ever need anything more complex we can change later. Instead of `appId` it should include a prefix so I think that something like `app.fybrik.io/tracing-uuid` might be more appropriate.

**Reply by roee88** [reply]
@froesef @shlomitk1 @revit13 any comments on the approach?

**Reply by shlomitk1** [reply]
Just to verify, if fybrikapplication has this tracing-uuid label, the label value will be used in all manager and connectors logs, and propagated to modules as other application labels. We'll need then to pass this value to the connectors as well, right?

**Reply by roee88** [reply]
Correct.

**Reply by shlomitk1** [reply]
I'm fine with this approach. Even if the tracing label is not set, it would be useful to use some id (app name, for instance). Otherwise, it is very hard to debug when more than one app are reconciled.

**Comment by laurieaw63** [comment]
@roee88 Long term the namespace cannot be UUID. It needs to be associated with the tenant. There will be one data access module namespace per tenant.  Short term we need to use UUID for the namespace until the full support for https://github.com/fybrik/fybrik/issues/793 is in place.

**Reply by roee88** [reply]
@lappanna wrote:
> we create fybrik application with a unique name like fybrik-<uuid>

Is that going to change as well?

**Comment by simanadler** [comment]
@lappanna and @roee88 What were the decisions for short term and long term?  Is there some function in utils.go or somewhere else that will generate the unique ID?

**Reply by shlomitk1** [reply]
@simanadler As I understand, the unique id will come from the app label, or generated (e.g. based on appName+namespace+creationStamp).

**Reply by simanadler** [reply]
For logging purposes we need the unique id to be unique across clusters and over time as well as per namespace.  We need to add the cluster as well.  I'm worried though that it will become quite long. 

I understand that kubernetes generates universally unique IDs for all objects, which are unique over time but not across clusters.  I propose the following:

Globally Unique ID for FybrikApplication
FybrikApplicationGUID := FybrikApplication.ObjectMeta.UID + "-" + FybrikApplication.ObjectMeta.ClusterName

Pass the FybrikApplicationGUID in the ObjectMeta.Labels of plotter, blueprint and modules so that it can be used by them when the create log entries.

What do you think?  @shlomitk1 @lappanna @roee88.

**Reply by shlomitk1** [reply]
I don't think cluster should be added. We want to track the same app over different clusters (modules running in the different clusters), thus it should be the same name.
uid of kubernetes is a good idea (it captures the app modifications as well).

**Reply by shlomitk1** [reply]
@simanadler ClusterName is defined in ObjectMeta but not set. Currently, we don't support generating FybrikApplications in worker clusters (not in the coordiantor cluster).

**Reply by simanadler** [reply]
@shlomitk1 OK, we can leave off the cluster for now.  However, we should create a function for getting the FybrikApplicationGUID so that in the future we can add the cluster if we need it without having to make broad changes in the code.

**Reply by simanadler** [reply]
@shlomitk1 Should the plotter owner now be the FybrikApplicationGUID instead of a NamespacedName?

**Reply by shlomitk1** [reply]
No, the communication between the app and the plotter is done based on name/namespace. Otherwise, you'd need a search mechanism to get fybrikapplication resource with the same label/uid.
UID is used for logging, it is not going to replace the way the resources are tracked.

**Reply by shlomitk1** [reply]
> 
> 
> @shlomitk1 OK, we can leave off the cluster for now. However, we should create a function for getting the FybrikApplicationGUID so that in the future we can add the cluster if we need it without having to make broad changes in the code.

Yes. The uid is determined either by using a dedicated app label or uid if no label is present. There will be a function for this.

**Reply by simanadler** [reply]
> No, the communication between the app and the plotter is done based on name/namespace. Otherwise, you'd need a search mechanism to get fybrikapplication resource with the same label/uid. UID is used for logging, it is not going to replace the way the resources are tracked.

Yes, that was my thinking as well.  Just wanted to make sure.

**Reply by simanadler** [reply]
I tried using the unique ID for a fybrikapplication  instance.  I tried to store  it as a label in fybrikapplication, plotter and blueprints.  However, the unique kubernetes ID does not pass the label validation.  As per [k8s documentation ](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/), 
'Valid label value:

    must be 63 characters or less (can be empty),
    unless empty, must begin and end with an alphanumeric character ([a-z0-9A-Z]),
    could contain dashes (-), underscores (_), dots (.), and alphanumerics between.
`
but the unique ID looks, for example, like this: "a037fc27-ebce-4200-8e2b-04cfabce88d9-"  

I guess we can pass it as part of specs of the plotter, blueprints, and modules instead of via a label unless you have any other idea.  @shlomitk1 @roee88

**Reply by roee88** [reply]
Options:
1. Set in an annotation instead
2. Use as field as suggested
3. Use namespace and name as identifying fields in logging because I still don't understand why a system that "create fybrik application with a unique name like `fybrik-<uuid>`" need an additional field. But I guess @lappanna and @laurieaw63 can explain why this isn't an option. Maybe the name has the same limits as a label?
